### PR TITLE
workbench.externalBrowser opens the selected browser properly

### LINKED ISSUE #230636: Windows: using `edge` as `workbench.externalBrowser` opens default browser too
Type: <b>Bug</b>

My default browser is Firefox.
I have `workbench.externalBrowser` set to `edge`
After running `npm run dev` on any web project and clicking the localhost link to open the website I would expect vscode to open the website in Edge only.
What actually happens is that VsCode open the website in both Firefox _and_ Edge

It does this will all extensions disabled.
It's very annoying.

It's a Vite React project I'm using, and in addition to opening browser tabs in both Edge and Firefox, VsCode is only actually establishing a browser link to the Firefox tab. So any compilation errors from Vite go to the Firefox console but not the Edge console. So despite having set `workbench.externalBrowser` to `edge`, the Edge browser tab that gets opened isn't even usable.

I see the same behaviour whether I run `npm run dev` in either the embedded cmd or Powershell terminals.

I also created a new create-react-app using the instructions from [here](https://code.visualstudio.com/docs/nodejs/reactjs-tutorial), and then manually added an Edge launch profile as described in that guide. This also had no effect and clicking the localhost link in the enbedded terminal just ignored the `workbench.externalBrowser` setting and opened the website in Firefox.

VS Code version: Code 1.94.0 (d78a74bcdfad14d5d3b1b782f87255d802b57511, 2024-10-02T13:08:12.626Z)
OS version: Windows_NT x64 10.0.22631
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|13th Gen Intel(R) Core(TM) i7-13700H (20 x 2918)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled<br>webnn: disabled_off|
|Load (avg)|undefined|
|Memory (System)|31.68GB (17.86GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details>Extensions: none

### PR DESCRIPTION
**Pull request by peterdanwan**
This fixes #230636.

Now, when the user tries to set the `workbench.externalBrowser` setting to a particular browser, it will not open 2 tabs (1 for the user's default browser, and the other tab for what the user put for the value of `workbench.externalBrowser`).

Below are some gifs that shows the effect of the changes:
open-link-fix-1

open-link-fix-2

### CODE REVIEW

**Comment by peterdanwan** [comment]
@microsoft-github-policy-service agree

File: src/vs/platform/native/electron-main/nativeHostMainService.ts [hunk 1]
```diff
@@ -522,32 +522,50 @@ export class NativeHostMainService extends Disposable implements INativeHostMain
...
+					command = `start ${browserName} "${url}"`;
+				} else if (isMacintosh) {
+					command = `open -a "${browserName}" "${url}"`;
+				} else {
+					command = `${browserName} "${url}"`;
+				}
+				await promisify(exec)(command);
+			} catch (fallbackError) {
+				this.logService.error(`Fallback method failed to open URL '${url}' using browser '${browserName}': ${fallbackError}`);
+			}
```
Comment (bpasero) [inline src/vs/platform/native/electron-main/nativeHostMainService.ts:568]: Why would we do this and not simply fallback to the `shell.openExternal`?

File: src/vs/platform/native/electron-main/nativeHostMainService.ts [hunk 2]
```diff
@@ -522,32 +522,50 @@ export class NativeHostMainService extends Disposable implements INativeHostMain
 			return shell.openExternal(url);
 		}
 
+		let browserName = configuredBrowser;
+		if (configuredBrowser.toLowerCase() === 'edge') {
+			browserName = isWindows ? 'msedge' : 'microsoft-edge';
+		}
```
Comment (bpasero) [inline src/vs/platform/native/electron-main/nativeHostMainService.ts:528]: Why is this needed?

File: src/vs/platform/native/electron-main/nativeHostMainService.ts [hunk 3]
```diff
@@ -522,32 +522,50 @@ export class NativeHostMainService extends Disposable implements INativeHostMain
...
 		try {
 			const { default: open } = await import('open');
 			const res = await open(url, {
 				app: {
-					// Use `open.apps` helper to allow cross-platform browser
-					// aliases to be looked up properly. Fallback to the
-					// configured value if not found.
-					name: Object.hasOwn(open.apps, configuredBrowser) ? open.apps[(configuredBrowser as keyof typeof open['apps'])] : configuredBrowser
-				}
+					name: browserName
```
Comment (bpasero) [inline src/vs/platform/native/electron-main/nativeHostMainService.ts:542]: I think we still need to lookup the browser name per platform using `open.apps` helper, no?

**Review by bpasero — CHANGES_REQUESTED** [review]
I think the gist of this change is to use `wait: true` for the browser execution which I find an interesting path forward because it could probably allow us to detect an abnormal process termination from a successful one. For example, when the child process terminates very quickly after having launched with a non-zero exit code, its likely that the launch failed. But it seems, using `wait: true` has implications for how the library can be used:

> On Windows, you have to explicitly specify an app for it to be able to wait.

File: src/vs/platform/native/electron-main/nativeHostMainService.ts (hunk 1 above)
Comment (peterdanwan) [reply src/vs/platform/native/electron-main/nativeHostMainService.ts:568]: Hi @bpasero! My mistake - I think I should have tried to make better use of the existing methods and the `open` library - since from what I now understand, the `open` library knows how to launch the different browser links properly without requiring me to make that manual if, else-if, and else block.

Comment (peterdanwan) [reply src/vs/platform/native/electron-main/nativeHostMainService.ts:568]: Do you by chance, have any further suggestions on how I should attempt to resolve this issue properly? I think I might need to do a little bit more research on the `openExternal` and `openExternalBrowser` methods + the `open` library to fix this while introducing the least amount of change.

**Comment by peterdanwan** [comment]
Hello @bpasero, I tried reverting the changes where I was manually checking which platform the user was on (reverted most changes back to how they were initially). 

In the call to the open function, I added the `wait: true` like what you suggested and it seemed to work on my side (I did the same type of tests I did in the gifs I sent earlier).

When you have the chance, could you let me know if this works better or if you would like me to revise anything else (e.g., I can squash the commits I made)?

Thank you!

**Comment by bpasero** [comment]
Thanks, but closing for now as a workaround is pushed and I am not sure the impact of `wait: true`.
